client/lib/utils/print-document.js

```diff
@@ -1,27 +1,68 @@
+import pdfSupport from 'lib/utils/pdf-support';
+
 let iframe = null;
 
 /**
- * Opens the native printing dialog to print the given URL.
+ * Loads the given URL in an invisible <iframe>
  * To do that, an invisible <iframe> is created, added to the current page, and "print()" is invoked
  * for just that iframe.
- * @param {string} url URL of the document to print
- * @returns {Promise} Promise that resolves when the document is loaded, and rejected if there's an error fetching it
+ * @param {string} url URL to load
+ * @returns {Promise} Promise that resolves when the iframe finished loading, rejects on error
  */
-export default ( url ) => {
+const loadDocumentInFrame = ( url ) => {
 	return new Promise( ( resolve, reject ) => {
 		if ( iframe ) {
 			document.body.removeChild( iframe );
 		}
 		iframe = document.createElement( 'iframe' );
 		iframe.src = url;
-		iframe.style.display = 'none';
+
+		// Note: Don't change this for "display: none" or it will stop working on MS Edge
+		iframe.style.position = 'fixed';
+		iframe.style.left = '-999px';
+
 		iframe.onload = () => {
-			iframe.contentWindow.print();
 			resolve();
 		};
 		iframe.onerror = ( error ) => {
 			reject( error );
 		};
+
 		document.body.appendChild( iframe );
 	} );
 };
+
+/**
+ * Opens the native printing dialog to print the given URL.
+ * Falls back to opening the PDF in a new tab if opening the printing dialog is not supported.
+ * @param {string} url URL of the document to print
+ * @param {boolean} isUserInteraction True if this code is being executed as a result of a direct user interaction,
+ * such as the click of a button. False otherwise (if this was triggered by an AJAX response, for example
+ * @returns {Promise} Promise that resolves with true if the printing dialog (or equivalent) was correctly
+ * invoked. It resolves with false if it's required to call this function again as a result of direct user interaction.
+ */
+export default ( url, isUserInteraction ) => {
+	switch ( pdfSupport() ) {
+		case 'native':
+			// Happy case where everything can happen automatically. Supported in Edge, Chrome and Safari
+			return loadDocumentInFrame( url )
+				.then( () => {
+					iframe.contentWindow.print();
+					return true;
+				} );
+
+		case 'addon':
+			// window.open will be blocked by the browser if this code isn't being executed from a direct user interaction
+			if ( isUserInteraction ) {
+				window.open( url );
+				return Promise.resolve( true );
+			}
+			return Promise.resolve( false );
```

Comment:
After the AJAX request for purchasing a label is completed, the function `printDocument(url, false)` is called, regardless of the browser capabilities. If the browser needs a direct user interaction, the Promise is resolved with `false`, that result is used [here](https://github.com/Automattic/woocommerce-services/pull/908/files/c224461151150ed30ac092db3f0e8888aec7a6ca#diff-d8b04016a3889c977c51e0ed0ae5a2f7R525), that code ultimately changes the `Purchasing...` button to `Print`. Since I had to explain how it works, that's a synthom that it's too complicated, I'll try to simplify.